src/coreclr/jit/hwintrinsicarm64.cpp

```diff
@@ -512,7 +558,7 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
         return gtNewScalarHWIntrinsicNode(TYP_VOID, intrinsic);
     }
 
-    assert(category != HW_Category_Scalar);
+    bool isScalar = (category == HW_Category_Scalar);
```

Comment:
can we be explicit about this?

`assert((category == HW_Category_Scalar) || (id == SaturatingDecrementBy8BitElementCount) || (id == SaturatingIncrementBy8BitElementCount))`

Reply:
At this point we've already potentially switched to the scalar variants. So the full assert would be:
```
    assert((category != HW_Category_Scalar) || (id == SaturatingDecrementBy8BitElementCount) || (id == SaturatingIncrementBy8BitElementCount)
            || (id == NI_Sve_SaturatingDecrementBy16BitElementCountScalar) || (id == NI_Sve_SaturatingDecrementBy32BitElementCountScalar)
            || (id == NI_Sve_SaturatingDecrementBy64BitElementCountScalar) || (id == NI_Sve_SaturatingIncrementBy16BitElementCountScalar)
            || (id == NI_Sve_SaturatingIncrementBy32BitElementCountScalar) || (id == NI_Sve_SaturatingIncrementBy64BitElementCountScalar));
```

Which feels excessive. Happy to switch if you still want.

Reply:
Fixed this by adding a DEBUG only bool.

Reply:
these are not `HW_Category_Scalar` and are not marked with `isValidScalarIntrinsic` ? does it work as expected?

Reply:
thanks for reminding me offline that they are `HW_Category_SIMD` and so should be fine.